### PR TITLE
[Core] OpenSSL 3 support

### LINKED ISSUE #1762: FS does not compile with openssl 3.x
**Describe the bug**
some functions are deprecated generating errors like
src/switch_apr.c: In function 'switch_md5':
src/switch_apr.c:1176:9: error: 'MD5_Init' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
 1176 |         MD5_Init(&md5_context);
      |         ^~~~~~~~
In file included from src/switch_apr.c:79:
/usr/include/openssl/md5.h:49:27: note: declared here
   49 | OSSL_DEPRECATEDIN_3_0 int MD5_Init(MD5_CTX *c);
      |                           ^~~~~~~~
src/switch_apr.c:1177:9: error: 'MD5_Update' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
 1177 |         MD5_Update(&md5_context, input, inputLen);
      |         ^~~~~~~~~~
/usr/include/openssl/md5.h:50:27: note: declared here
   50 | OSSL_DEPRECATEDIN_3_0 int MD5_Update(MD5_CTX *c, const void *data, size_t len);
      |                           ^~~~~~~~~~
src/switch_apr.c:1178:9: error: 'MD5_Final' is deprecated: Since OpenSSL 3.0 [-Werror=deprecated-declarations]
 1178 |         MD5_Final(digest, &md5_context);
      |         ^~~~~~~~~
/usr/include/openssl/md5.h:51:27: note: declared here
   51 | OSSL_DEPRECATEDIN_3_0 int MD5_Final(unsigned char *md, MD5_CTX *c);

**To Reproduce**
compile FS with openssl 3.x

### PR DESCRIPTION
**Pull request by jpfox156**
Low-level openssl api for accessing md5 was deprecated in openssl3.0, leading to errors during compile of freeswitch. 
Update switch_md5 method to use the openssl high-level api. 
syntax plagiarized from: 
https://stackoverflow.com/questions/69806220/advice-needed-for-migration-of-low-level-openssl-api-to-high-level-openssl-apis

### CODE REVIEW

**Comment by ar45** [comment]
@jpfox156 Can you rebase this?

**Review by ar45 — APPROVED** [review]

**Comment by ar45** [comment]
@andywolk Can you please review this?

File: src/switch_apr.c [hunk 1]
```diff
@@ -1174,12 +1183,22 @@ SWITCH_DECLARE(switch_status_t) switch_md5(unsigned char digest[SWITCH_MD5_DIGES
...
-	MD5_Final(digest, &md5_context);
-
+	#if OPENSSL_VERSION_NUMBER < 0x30000000
+		MD5_CTX md5_context;
+		MD5_Init(&md5_context);
+		MD5_Update(&md5_context, input, inputLen);
+		MD5_Final(digest, &md5_context);
+	#else
+		EVP_MD_CTX *md5_context;
+		// MD5_Init
```
Comment (andywolk) [inline src/switch_apr.c:1193]: Please use `/* */` comment style

File: src/switch_rtp.c [hunk 2]
```diff
@@ -3716,13 +3720,20 @@ SWITCH_DECLARE(switch_status_t) switch_rtp_add_dtls(switch_rtp_t *rtp_session, d
...
 	if (dh) {
 		SSL_CTX_set_tmp_dh(dtls->ssl_ctx, dh);
 		DH_free(dh);
 	}
-
+#else 
+	if((dh_pk = PEM_read_bio_PrivateKey(bio, NULL, NULL, NULL)) != NULL) {
+		SSL_CTX_set0_tmp_dh_pkey(dtls->ssl_ctx, dh_pk);
+		EVP_PKEY_free(dh_pk);
+	}
```
Comment (andywolk) [inline src/switch_rtp.c:3734]: A new line after an `if` block is required.

**Review by andywolk — CHANGES_REQUESTED** [review]

File: src/switch_core_cert.c [hunk 3]
```diff
@@ -432,7 +434,23 @@ static int mkcert(X509 **x509p, EVP_PKEY **pkeyp, int bits, int serial, int days
 		x = *x509p;
 	}
 
-#if OPENSSL_VERSION_NUMBER >= 0x10100000
+#if OPENSSL_VERSION_NUMBER >= 0x30000000
+	{
+		EVP_PKEY_CTX *ctx;
+		ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
```
Comment (andywolk) [inline src/switch_core_cert.c:440]: A new line after declarations is required.

File: src/switch_core_cert.c [hunk 4]
```diff
@@ -432,7 +434,23 @@ static int mkcert(X509 **x509p, EVP_PKEY **pkeyp, int bits, int serial, int days
...
+		/* Setup the key context */
+		if ((!ctx) || (EVP_PKEY_keygen_init(ctx) <= 0) || (EVP_PKEY_CTX_set_rsa_keygen_bits(ctx, bits) <= 0)) {
+			abort();
+			goto err;
+		}
+		/* Generate key */
```
Comment (andywolk) [inline src/switch_core_cert.c:446]: A new line after an if block

File: src/switch_core_cert.c [hunk 5]
```diff
@@ -432,7 +434,23 @@ static int mkcert(X509 **x509p, EVP_PKEY **pkeyp, int bits, int serial, int days
...
+		if (EVP_PKEY_generate(ctx, &pk) <= 0) {
+			abort();
+			goto err;
+		}
+		EVP_PKEY_CTX_free(ctx);
```
Comment (andywolk) [inline src/switch_core_cert.c:451]: A new line after an if block

File: src/switch_rtp.c [hunk 6]
```diff
@@ -3716,13 +3720,22 @@ SWITCH_DECLARE(switch_status_t) switch_rtp_add_dtls(switch_rtp_t *rtp_session, d
...
 	bio = BIO_new_file(dtls->pem, "r");
+#if OPENSSL_VERSION_NUMBER < 0x30000000
 	dh = PEM_read_bio_DHparams(bio, NULL, NULL, NULL);
 	BIO_free(bio);
 	if (dh) {
 		SSL_CTX_set_tmp_dh(dtls->ssl_ctx, dh);
 		DH_free(dh);
 	}
-
+
```
Comment (andywolk) [inline src/switch_rtp.c:3730]: We can skip that new line

**Comment by daniel-lucio** [comment]
What is preventing this to be accepted?

**Comment by xadhoom** [comment]
With the current changes FS may segfault, depending on openssl 3 config.

The crash happens because the call to `mkcert` in `switch_core_gen_certs` is not checked on return value, and if the return value is 0 (which means that certs/keys gen have failed) the pointers are not set but still passed to `PEM_write_PrivateKey` which segfaults if `pkey` is not set.

The reason why `mkcert` fails is because (at least on rhel9) `EVP_sha1` is deprecated which fails `X509_sign`.

So I suggest two addtions:
- check retval for `mkcert` in `switch_core_gen_certs`, something like:
```patch
-	mkcert(&x509, &pkey, 4096, 0, 36500);
+	if (!mkcert(&x509, &pkey, 4096, 0, 36500)) {
+		switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_ERROR, "Certificate generation failed\n");
+		abort();
+	}

```

then switch to `EVP_sha256` when signing the certificate in `mkcert`, something like:
```patch
 	 */
 	X509_set_issuer_name(x, name);
 
+#if OPENSSL_VERSION_NUMBER >= 0x30000000
+	if (!X509_sign(x, pk, EVP_sha256()))
+#else
 	if (!X509_sign(x, pk, EVP_sha1()))
+#endif
 		goto err;
 
 	*x509p = x;
 	*pkeyp = pk;
 	return(1);
  err:
+	ERR_print_errors_fp(stdout);
 	return(0);
 }
 
```

Note that I've added also a call to `ERR_print_errors_fp` on the `err` label in order to help diagnose.

**Review by mjerris — APPROVED** [review]

**Pull request merged**